tests/backend/checkmach/test_assume_fail.ml

```diff
@@ -10,3 +10,42 @@ let[@zero_alloc assume][@inline always] bar' x =
   (x+1,x)
 
 let[@zero_alloc strict] foo' x = bar' x
+
+let[@inline always] test46 x = if x > 0 then failwith (Printf.sprintf "%d" x) else (x,x)
+
+let[@zero_alloc strict] test48 x =
+  (test46[@zero_alloc assume never_returns_normally]) x
+
+(* Perhaps confusingly, never_returns_normally works on allocations not only
+   on calls. This is needed for analysis to give the same results regardess
+   of inlining of expressions annotated with "assume". See test50-53 below.  *)
```

Comment:
> Also, perhaps there should be a test showing how this works in the context of a strict check. For example, if I have
> 
> ```ocaml
> let test1 x = x
> 
> let[@zero_alloc strict] test2 x =
>   let z = (test1[@zero_alloc assume never_returns_normally]) x in z, z
> ```
> 
> Does the check pass? I think it should, since the allocation is assumed to be unreachable.

When `test1` is not inlined, this test fails because `test1` may raise and may allocate before it raises. The spec `assume never_returns_normally` doesn't guarantee anything about executions of `test1` that return via an exception. In terms of abstract values above, `exn` component of `assume never_returns_normally` is `Top`.  The error message we get in this case explains reasonably well why the check fails.

However, when `test1` is inlined, we get a different error message that may be confusing to the user: "Unexpected allocation of 24 bytes" pointing to the pair `z,z` in `test2`. Here, after the function is inlined, we know that it returns normally and the "assume" annotation is not taken into account.  This is annoying, but not so bad because the analysis outcome is FAIL in both cases. 

If we also add  "strict" keyword to this "assume" annotation, the check will PASS when `test1` function is not inlined, but fail when `test` is not inlined. This is hard to fix because after inlining and optimizations there is nothing left to attach the annotation to.

Reply:
Perfect - thanks!